File: golang/_0025_Reverse_Nodes_in_k-Group/main.go
package main

import "leetcode/golang/common"

type ListNode = common.ListNode

func reverseKGroup(head *ListNode, k int) *ListNode {
	if head == nil || head.Next == nil || k <= 1 {
		return head
	}

	dummy := new(ListNode)
	tail := dummy
	for head != nil {
		pre := head
		i := 0
		for i < k-1 && pre.Next != nil {
			i++
			pre = pre.Next
		}

		if i < k-1 {
			tail.Next = head
			break
		}

		nextHead := pre.Next
		pre.Next = nil
		tail.Next = reverse(head)
		for tail.Next != nil {
			tail = tail.Next
		}

		head = nextHead
	}

	return dummy.Next
}

func reverse(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	dummy := new(ListNode)
	for head != nil {
		next := head.Next
		head.Next = dummy.Next
		dummy.Next = head
		head = next
	}

	return dummy.Next
}

File: python3/_0025_Reverse_Nodes_in_k-Group/main.py
from python3.common.define import ListNode


class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next or k <= 1:
            return head

        dummy = tail = ListNode(0)
        while head:
            cnt = 1
            cur = head
            while cur and cnt < k:
                cnt += 1
                cur = cur.next

            if cur:
                t = cur.next

                cur.next = None
                while head:
                    tt = head.next
                    head.next = tail.next
                    tail.next = head
                    head = tt

                while tail.next:
                    tail = tail.next

                head = t
            else:
                tail.next = head
                head = None

        return dummy.next
